cmake: Toolchain abstraction: C++ (part 3)

mped-oticon opened the pull request:
(Part 3 of N. Based on top of Part 2. Part 2 in https://github.com/zephyrproject-rtos/zephyr/pull/12859)

Introduce toolchain_cc_cpp macro.

Move the following into the toolchain_cc_cpp macro:
 * Dialect/ISO version
 * RTTI
 * Exceptions

No functional change expected.

Clang's C++ flags are compatible with gcc, and are thus
inherited.

This is motivated by the wish to abstract Zephyr's usage of toolchains,
permitting easier porting to other (commercial) toolchains.

Signed-off-by: Mark Ruvald Pedersen <mped@oticon.com>

-------------

Part of #16031

SebastianBoe reviewed (changes requested):
This is blocked by, and will be reviewed after

https://github.com/zephyrproject-rtos/zephyr/pull/12859#discussion_r252668487

SebastianBoe commented on CMakeLists.txt:
```diff
@@ -157,27 +145,10 @@ zephyr_compile_options(
-  $<$<COMPILE_LANGUAGE:CXX>:-std=${STD_CPP_DIALECT}>
-  $<$<COMPILE_LANGUAGE:CXX>:-fcheck-new>
-  $<$<COMPILE_LANGUAGE:CXX>:-ffunction-sections>
-  $<$<COMPILE_LANGUAGE:CXX>:-fdata-sections>
```
This line and the above is redundant with

zephyr_cc_option(-ffunction-sections -fdata-sections)

from arch/common/CMakeLists.txt

and can be safely removed, instead of ported.

(I ran some tests to confirm, the ordering of flags is changed, but the set of flags remains unchanged)

As an aside, you could move

zephyr_cc_option(-ffunction-sections -fdata-sections)

to the toplevel CMakeLists.txt if you wish.

mped-oticon replied:
Fixed in separate commit.

SebastianBoe commented on cmake/compiler/gcc/target_cpp.cmake:
```diff
@@ -0,0 +1,51 @@
+    zephyr_compile_options(
+      $<$<COMPILE_LANGUAGE:CXX>:-fno-exceptions>
+    )
+  endif()
```
Same feedback for exceptions, as for https://github.com/zephyrproject-rtos/zephyr/pull/12413#discussion_r249423435

mped-oticon replied:
@SebastianBoe I've followed the same stencil from Part 2.

mped-oticon committed:
cmake: Toolchain abstraction: C++

Introduce toolchain_cc_cpp_*-family of macros.

Move the following into the toolchain_cc_cpp_* macros:
 * Common base set of flags
 * C++ standard version
 * RTTI
 * Exceptions

These macros must be implemented by every compiler port.
These macros set the respective flags, but leaves logic and control to
the root CMakeLists.txt file.

No functional change expected.

Clang's C++ flags are compatible with gcc, and are thus
inherited.

This is motivated by the wish to abstract Zephyr's usage of toolchains,
permitting easier porting to other (commercial) toolchains.

Signed-off-by: Mark Ruvald Pedersen <mped@oticon.com>

mped-oticon committed:
cmake: remove redundant compiler flags for C++

The compiler flags -ffunction-sections and -fdata-sections are already
universally set from arch/common/CMakeLists.txt, so we shall not set
them again for C++ only.

Signed-off-by: Mark Ruvald Pedersen <mped@oticon.com>

mped-oticon commented:
Please re-review.

mped-oticon commented:
CI failed because I thought it _would just work_ :)
It because we do checking on command line arguments, and for the non-98 versions we include a second argument, space delimited. E.g. '-foo -bar' is treated as one argument.

I'm fixing this by making it into a list. Let me know if you have other ideas.

mped-oticon commented:
Pushed the list fix.
I have verified this works using strace, observing all intended flags are in fact passed:
```
$ strace -f -e execve time sanitycheck -s tests/application_development/cpp/test 2>strace.log
```
